### ISSUE TITLE
Breaking change for MVC action filters in version 8.2+

### ISSUE THREAD

**Issue by psilver-schumacher** [umbraco/Umbraco-CMS#6813]
After upgrading to Umbraco 8.2 Any unhandled exceptions instead of showing the 500 error handler specified by the customerrors section for asp.net OR the default IIS error handler will instead simply produce an internal error state that eventually causes  the application pool to shut down and no response is returned to the client.

Reproduction - under Windows Server 2012, IIS 8 (also under Windows Server 2016 same issue)
------------

1. Have your custom errors set to On, specify an error handler for 500 error code
2. Throw an exception in a rendermvc controller
3. Have tryskipiiserrors = true in the umbraco settings routing entry
4. have rapid fail protection enabled on your IIS app pool

Once the exception is thrown you simply get a connection reset in your browser instead of the expected 500 error page.

Verified that if we rolled back to Umbraco 8.1.4 this is no longer a problem and the 500 error page displays as expected so our configuration was not the issue.

Here are the two relevant lines from the serilog logs set to full DEBUG detail level.  You can see that the app hits the Umbraco error handler, is not a redirect error, and then Umbraco immediately restarts.  I removed the server name as it is not relevant here.

{"@t":"2019-10-22T15:21:09.6875635Z","@mt":"Response status: Redirect={Redirect}, Is404={Is404}, StatusCode={ResponseStatusCode}","@l":"Debug","Redirect":"none","Is404":"false","ResponseStatusCode":0,"SourceContext":"Umbraco.Web.UmbracoModule","ProcessId":5780,"ProcessName":"w3wp","ThreadId":8,"AppDomainId":4,"AppDomainAppId":"LMW3SVC16ROOT","MachineName":"<redacted>","Log4NetLevel":"DEBUG","HttpRequestNumber":8,"HttpRequestId":"eecb528c-15a2-46a5-9ed4-d30c66c5f2aa"}
{"@t":"2019-10-22T15:21:14.3752236Z","@mt":"{StartMessage} [Timing {TimingId}]","@l":"Debug","StartMessage":"Render Macro: Navigation, type: PartialView, cache: 0","TimingId":"d433539","SourceContext":"Umbraco.Web.Macros.MacroRenderer","ProcessId":5780,"ProcessName":"w3wp","ThreadId":8,"AppDomainId":4,"AppDomainAppId":"LMW3SVC16ROOT","MachineName":"<redacted>","Log4NetLevel":"DEBUG","HttpSessionId":"wf4rjyveqbfmeu2dix0uqr5e","HttpRequestNumber":8,"HttpRequestId":"eecb528c-15a2-46a5-9ed4-d30c66c5f2aa"}

Interestingly this issue does not happen in local debug mode on Windows 10 with IIS Express.  Running there with customerrors set to off we get the expected Yellow screen of death.  I tried to reproduce it under those conditions and could not.  I have not tried using actual IIS locally but if I have time I will.

**Comment by psilver-schumacher** [comment]
Update:

After local testing I can verify that
1. This issue only happens under IIS and NOT IIS Express
2. The issue does not happen under 8.1.x only 8.2.x, so the code in question DID work perfectly in 8.1
3. I was able to pin point the problem specifically to MVC action filters.

Normally I would just use LightInject's native action filter support to wire in MVC action filters but that was not possible from what I tried in Umbraco's environment.  Instead I used the accepted technique of setting up my own GlobalFilterProvider (source attached) which is not doing anything special.  If I disable all my action filters in 8.2 the problem stops - exceptions are handled normally again.  Not sure what changed in this version or if there is some workaround so I can still use action filters? 

 public class GlobalFilterProvider : IFilterProvider
    {
        private readonly IDependencyResolver dependencyResolver;

        public GlobalFilterProvider(IDependencyResolver dependencyResolver)
        {
            this.dependencyResolver = dependencyResolver;
        }

        public IEnumerable<Filter> GetFilters(ControllerContext controllerContext, ActionDescriptor actionDescriptor)
        {
            foreach (var filter in this.dependencyResolver.GetServices<IActionFilter>())
            {
                yield return new Filter(filter, FilterScope.Global, order: null);
            }
            foreach (var filter in this.dependencyResolver.GetServices<IAuthorizationFilter>())
            {
                yield return new Filter(filter, FilterScope.Global, order: null);
            }
            foreach (var filter in this.dependencyResolver.GetServices<IExceptionFilter>())
            {
                yield return new Filter(filter, FilterScope.Global, order: null);
            }
            foreach (var filter in this.dependencyResolver.GetServices<IResultFilter>())
            {
                yield return new Filter(filter, FilterScope.Global, order: null);
            }
            foreach (var filter in this.dependencyResolver.GetServices<System.Web.Mvc.Filters.IAuthenticationFilter>())
            {
                yield return new Filter(filter, FilterScope.Global, order: null);
            }
        }
    }

And the corresponding composer component

 public class FilterComponent : IComponent
    {
        public void Initialize()
        {
            // Register the filter provider with MVC - normally LightInject MVC allows property injection
            //  of MVC attributes but with Umbraco this doesn't seem to work in testing.
            //  Doing it this way will allow us to constructor inject our action filters which LightInject doesn't support
            //  anyhow.
            FilterProviders.Providers.Insert(0, new GlobalFilterProvider(DependencyResolver.Current));
        }
        
        public void Terminate()
        {
            // Does nothing
        }
    }

**Comment by psilver-schumacher** [comment]
For the sake of anyone else who has to deal with the same issues - we had to use action filters to get us some context data and other "global" functionality across the entire site.  After a lot of experimenting, it seems that only certain types of action filters cause the breaking behavior where it crashes the app pool.

Specifically you CAN inject

IActionFilter (which is good that is the main one I need)
IExceptionFilter (which is also good for global error handling)

You can NOT inject any of these without causing the crashing behavior:
IResultFilter
IAuthenticationFilter
IAuthorizationFilter

So if I modify the pasted source code from the prior post for 8.2 - this will work - hopefully future versions don't break it as action filters are a good way to do global things.

The now working GlobalFilterProvider code for 8.2:

```
public class GlobalFilterProvider : IFilterProvider
    {
        private readonly IDependencyResolver dependencyResolver;

        public GlobalFilterProvider(IDependencyResolver dependencyResolver)
        {
            this.dependencyResolver = dependencyResolver;
        }

        public IEnumerable<Filter> GetFilters(ControllerContext controllerContext, ActionDescriptor actionDescriptor)
        {
            foreach (var filter in this.dependencyResolver.GetServices<IActionFilter>())
            {
                yield return new Filter(filter, FilterScope.Global, order: null);
            }
            foreach (var filter in this.dependencyResolver.GetServices<IExceptionFilter>())
            {
                yield return new Filter(filter, FilterScope.Global, order: null);
            }
        }
    }
```

**Comment by nul800sebastiaan** [comment]
Hey there @psilver-schumacher - this all sounds very very weird and it's hard to wrap my head around. I've asked around and everybody is quite puzzled. IIS Express and IIS are really very much the same so if the behavior differs between the two then there may be a bigger problem that Microsoft needs to look into.

I would advise you to try this out on a brand new 8.2.0 install with the custom code that seems to be problematic for you. If you can still reproduce this on a clean install then it is something we could look into but to the best of my (and my team's knowledge), there's nothing really changed in the areas that you're touching.

**Comment by psilver-schumacher** [comment]
Thank you for checking on this.  I created a clean bare bones project with umbraco 8.2 and put the minimal code in place to apply the action filters and attempt to recreate the same scenario.  Thus far I have not been able to.  It may be something with the other dependencies that got updated in the upgrade process but I think I can rule out core Umbraco as the issue at this point.  

I will close this issue as not a bug at this point - If I do discover any more about this as to the cause I will comment here in case someone else runs into this in the future.

Thanks!  Looking forward to the 8.3 update.
